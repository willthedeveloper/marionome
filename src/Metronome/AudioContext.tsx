import React from "react";

interface AudioContextProvider {
	audioCtx: AudioContext | undefined;
	createAudioCtx: Function;
}

const defaultValue: AudioContextProvider = {
	audioCtx: undefined,
	createAudioCtx: () => {}
};

const AudioContext = React.createContext(defaultValue);

export const AudioProvider = AudioContext.Provider;
export const AudioConsumer = AudioContext.Consumer;

export default AudioContext;
