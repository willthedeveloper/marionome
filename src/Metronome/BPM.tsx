import React from "react";
import InputRange from "./InputRange";
import style from "../styles/bpm.module.scss";
import layout from "../styles/layout.module.scss";
import cx from "classnames";

type BPMProps = {
	className?: string;
	value: number;
	min: number;
	max: number;
	handleChange: Function;
};

function BPM({ className, value, min, max, handleChange }: BPMProps) {
	return (
		<div className={cx(layout.flexCenter)}>
			<span className={style.numDisplay}>{value}</span>
			<InputRange
				value={value}
				min={min}
				max={max}
				onChange={(e) => handleChange(e)}
				ariaLabelForHandle='bpm-slider'
			/>
		</div>
	);
}

export default BPM;
