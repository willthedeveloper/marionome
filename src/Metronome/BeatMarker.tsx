import * as React from "react";
import styles from "../styles/beat.module.scss";

function BeatMarker() {
	return (
		<div className={styles.beatMarker}></div>
	);
}

export default BeatMarker;
