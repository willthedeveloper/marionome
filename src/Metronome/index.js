import layout from "../styles/layout.module.scss";
import InputRange from "./InputRange";
import DotContainer from "./DotContainer";
import cx from "classnames";

function Metronome() {
	return (
		<div className={layout.grid}>
			<div className={layout.colMid}>
				<InputRange className={layout.padVert} />
				<DotContainer />
			</div>
		</div>
	);
}

export default Metronome;
