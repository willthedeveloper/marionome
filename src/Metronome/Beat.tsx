import * as React from "react";
import InputRange from "./InputRange";
import BeatMarker from "./BeatMarker";
import cx from "classnames";

type BeatProps = {
    className?: string;
}

function Beat({className}: BeatProps) {
	return (
		<div className={cx("beat", className)}>
            <BeatMarker className={className} />
			<InputRange orientation='vertical' />
		</div>
	);
}

export default Beat;
