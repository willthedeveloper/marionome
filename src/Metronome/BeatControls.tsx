import React from "react";
import layout from "../styles/layout.module.scss";
import cx from "classnames";
import { BeatState } from "./Beat";
import { ImMinus, ImPlus } from "react-icons/im";
import color from "../styles/color";

type BeatControlsProps = {
	className?: string;
	beats: BeatState[];
	setBeats: Function;
};

function BeatControls({ className, beats, setBeats }: BeatControlsProps) {
	const handleAddBeat = () => {
		let newBeats = beats.slice();
		newBeats.push({ volume: 50 });
		setBeats(newBeats);
	};

	const handleRemoveBeat = () => {
		// Minimum beats = 1
		if (beats.length <= 1) {
			return;
		}
		setBeats(beats.slice(0, beats.length - 1));
	};

	let btnStyle = {
		backgroundColor: color.primaryLightest,
		color: color.primaryDark
	};

	return (
		<div className={cx(layout.flexCenter)}>
			<button aria-label='remove-beat' onClick={handleRemoveBeat} style={btnStyle}>
				<ImMinus />
			</button>
			<button aria-label='add-beat' onClick={handleAddBeat} style={btnStyle}>
				<ImPlus />
			</button>
		</div>
	);
}

export default BeatControls;
