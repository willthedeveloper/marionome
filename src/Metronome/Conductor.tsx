import React, { useState } from "react";
import Beat, { BeatState } from "./Beat";
import layout from "../styles/layout.module.scss";
import cx from "classnames";

type ConductorProps = {
	className?: string;
};

function Conductor({ className }: ConductorProps) {
	let defaultBeats: BeatState[] = [
		{ volume: 100, on: true },
		{ volume: 50, on: false },
		{ volume: 25, on: false },
		{ volume: 50, on: false }
	];

	let [beats, setBeats] = useState(defaultBeats);

	const handleVolUpdate = (index: number, val: number) => {
		let newBeats = [...beats];
		newBeats[index].volume = val;
		setBeats(newBeats);
	};

	const handleAddBeat = () => {
		let newBeats = beats.slice();
		newBeats.push({ volume: 50, on: false });
		setBeats(newBeats);
	};

	const handleRemoveBeat = () => {
		setBeats(beats.slice(0, beats.length - 1));
	};

	return (
		<div className={layout.row}>
			<div className={cx(layout.flexCenter, layout.left)}>
				{beats.map((beat, key) => (
					<Beat
						volume={beat.volume}
						on={beat.on}
						key={key}
						onVolumeUpdate={(val) => handleVolUpdate(key, val)}
					/>
				))}
			</div>
			<div className={cx(layout.flexCenter, layout.right)}>
				<button aria-label='add-beat' onClick={handleAddBeat}>
					+ beat
				</button>
				<button aria-label='remove-beat' onClick={handleRemoveBeat}>
					- beat
				</button>
			</div>
		</div>
	);
}

export default Conductor;
