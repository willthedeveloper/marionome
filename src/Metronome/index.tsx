import React, { useState, useEffect } from "react";
import layout from "../styles/layout.module.scss";
import BPM from "./BPM";
import Conductor from "./Conductor";
import PlayButton from "./PlayButton";
import { AudioProvider } from "./AudioContext";
import { BeatState } from "./Beat";

function createContext(): AudioContext {
	const AudioCtx = window.AudioContext;
	const audioCtx = new AudioCtx();
	return audioCtx;
}

let defaultBeats: BeatState[] = [{ volume: 100 }, { volume: 50 }, { volume: 25 }, { volume: 50 }];

let settings = {
	isPlaying: false,
	bpm: 88,
	beats: defaultBeats,
	currentBeat: -1
};

let audioCtx: AudioContext | undefined = undefined;
let nextBeatTime: number = 0;
let timerID: NodeJS.Timeout;
let lookahead = 25.0; // How frequently to call scheduling function (in milliseconds)
let scheduleAheadTime = 0.1; // How far ahead to schedule audio (sec)
const notesInQueue: { note: number; time: number }[] = [];

let audioBuffers: [] | AudioBuffer[] = [];

const getNextNoteTime = (currTime: number, bpm: number) => {
	const secondsPerBeat = 60.0 / bpm;
	return currTime + secondsPerBeat;
};

function scheduleNote(currentBeat: number, volume: number, time: number) {
	notesInQueue.push({ note: currentBeat, time: time });
	if (currentBeat === 0) {
		playSoundAtTime(audioBuffers[1], volume, time); // beat sound
	} else {
		playSoundAtTime(audioBuffers[0], volume, time); // accent sound
	}
}

const playSoundAtTime = (buffer: AudioBuffer | null, volume: number, time: number) => {
	const sampleSource = audioCtx!.createBufferSource();
	const gainNode = audioCtx!.createGain();
	sampleSource.buffer = buffer;
	gainNode.gain.value = volume / 100;
	
	sampleSource.connect(gainNode);
	gainNode.connect(audioCtx!.destination);
	sampleSource.start(time);
};

async function setupSamples(audioContext: AudioContext): Promise<AudioBuffer[]> {
	const samples = [`click.wav`, `accent.wav`];
	const audioBuffers = await Promise.all(
		samples.map(async (sample) => {
			const response = await fetch(`${process.env.PUBLIC_URL}/sounds/${sample}`);
			const arrayBuffer = await response.arrayBuffer();
			return audioContext.decodeAudioData(arrayBuffer);
		})
	);
	return audioBuffers;
}

const loadSamples = async () => {
	audioCtx = createContext();
	audioBuffers = await setupSamples(audioCtx);
};

function nextBeat(prevBeat: number, beats: BeatState[]): number {
	// Advance the beat number, wrap to zero
	if (prevBeat >= beats.length - 1 || prevBeat === -1) {
		return 0
	} else {
		return prevBeat + 1;
	}
}

function Metronome() {
	const [isPlaying, setPlaying] = useState(settings.isPlaying);
	const [bpm, setBPM] = useState(settings.bpm);
	const [currentBeat, setCurrentBeat] = useState(settings.currentBeat);
	const [beats, setBeats] = useState(settings.beats);

	const start = async () => {
		// Initialize audio if needed
		if (!audioCtx) {
			await loadSamples();
		} // check if context is in suspended state (autoplay policy)
		else if (audioCtx.state === "suspended") {
			audioCtx.resume();
		}

		nextBeatTime = audioCtx!.currentTime;
	};

	useEffect(() => {
		function scheduler() {
			const currentTime = audioCtx!.currentTime;
			// While there are notes that will need to play before the next interval, schedule them and advance the pointer.
			setCurrentBeat((prevBeat) => {
				while (nextBeatTime < currentTime + scheduleAheadTime) {
					const beatIndex = nextBeat(prevBeat, beats);
					const volume = beats[beatIndex].volume;

					scheduleNote(beatIndex, volume, nextBeatTime);
					nextBeatTime = getNextNoteTime(currentTime, bpm);
					return prevBeat;
				}
				return prevBeat;
			});

			timerID = setTimeout(scheduler, lookahead);
		}

		function pollForBeat() {
			const currentTime = audioCtx!.currentTime;

			// Fires when there are notes that need to be played
			while (notesInQueue.length && notesInQueue[0].time < currentTime) {
				setCurrentBeat(notesInQueue[0].note);
				notesInQueue.splice(0, 1); // remove note from queue
			}

			requestAnimationFrame(pollForBeat);
		}

		if (isPlaying) {
			scheduler();
			requestAnimationFrame(pollForBeat);
		}

		return () => {
			clearTimeout(timerID);
		};
	}, [isPlaying, beats, bpm]);

	const handlePlayToggle = async () => {
		if (isPlaying) {
			setCurrentBeat(-1);
		} else {
			await start();
		}
		setPlaying(!isPlaying);
	};

	return (
		<div className={layout.container}>
			<AudioProvider
				value={{
					audioCtx: undefined,
					createAudioCtx: createContext
				}}
			>
				<BPM value={bpm} min={40} max={240} handleChange={setBPM} />
				<Conductor beats={beats} currentBeat={currentBeat} setBeats={setBeats} />
				<PlayButton isPlaying={isPlaying} handleToggle={handlePlayToggle} />
			</AudioProvider>
		</div>
	);
}

export default Metronome;
