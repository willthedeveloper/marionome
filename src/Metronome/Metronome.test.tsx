import React from "react";
import { render, cleanup } from "@testing-library/react";
import Metronome from ".";

afterEach(cleanup);

it("should take a snapshot of the whole app", () => {
	const { asFragment } = render(<Metronome />);
	expect(asFragment(<Metronome />)).toMatchSnapshot();
});
