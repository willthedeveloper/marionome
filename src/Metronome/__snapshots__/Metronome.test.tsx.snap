// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot of the whole app 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="flexCenter left"
      >
        <span
          class="numDisplay"
        >
          88
        </span>
        <div
          class="rc-slider"
        >
          <div
            class="rc-slider-rail"
            style="background-color: rgb(241, 250, 238); width: 100%; height: 100%; border-radius: 0;"
          />
          <div
            class="rc-slider-track"
            style="background-color: rgb(168, 218, 220); width: 24%; height: 100%; left: 0%; border-radius: 0;"
          />
          <div
            class="rc-slider-step"
          />
          <div
            aria-disabled="false"
            aria-label="bpm-slider"
            aria-valuemax="240"
            aria-valuemin="40"
            aria-valuenow="88"
            class="rc-slider-handle"
            role="slider"
            style="height: 0px; border-radius: 0; margin: 0px; left: 24%; transform: translateX(-50%);"
            tabindex="0"
          />
          <div
            class="rc-slider-mark"
          />
        </div>
      </div>
      <div
        class="flexCenter right"
      >
        <button
          aria-label="increase-4-bpm"
        >
          +4 bpm
        </button>
        <button
          aria-label="decrease-4-bpm"
        >
          -4 bpm
        </button>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="flexCenter left"
      >
        <div
          class="beat"
        >
          <div
            class="rc-slider rc-slider-vertical"
          >
            <div
              class="rc-slider-rail"
              style="background-color: rgb(69, 123, 157); width: 100%; height: 100%; border-radius: 0;"
            />
            <div
              class="rc-slider-track"
              style="background-color: rgb(29, 53, 87); width: 100%; height: 100%; left: 0px; border-radius: 0; bottom: 0%;"
            />
            <div
              class="rc-slider-step"
            />
            <div
              aria-disabled="false"
              aria-label="beat-slider"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="100"
              class="rc-slider-handle"
              role="slider"
              style="height: 0px; border-radius: 0; margin: 0px; bottom: 100%; transform: translateY(+50%);"
              tabindex="0"
            />
            <div
              class="rc-slider-mark"
            />
          </div>
        </div>
        <div
          class="beat"
        >
          <div
            class="rc-slider rc-slider-vertical"
          >
            <div
              class="rc-slider-rail"
              style="background-color: rgb(241, 250, 238); width: 100%; height: 100%; border-radius: 0;"
            />
            <div
              class="rc-slider-track"
              style="background-color: rgb(168, 218, 220); width: 100%; height: 50%; left: 0px; border-radius: 0; bottom: 0%;"
            />
            <div
              class="rc-slider-step"
            />
            <div
              aria-disabled="false"
              aria-label="beat-slider"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="50"
              class="rc-slider-handle"
              role="slider"
              style="height: 0px; border-radius: 0; margin: 0px; bottom: 50%; transform: translateY(+50%);"
              tabindex="0"
            />
            <div
              class="rc-slider-mark"
            />
          </div>
        </div>
        <div
          class="beat"
        >
          <div
            class="rc-slider rc-slider-vertical"
          >
            <div
              class="rc-slider-rail"
              style="background-color: rgb(241, 250, 238); width: 100%; height: 100%; border-radius: 0;"
            />
            <div
              class="rc-slider-track"
              style="background-color: rgb(168, 218, 220); width: 100%; height: 25%; left: 0px; border-radius: 0; bottom: 0%;"
            />
            <div
              class="rc-slider-step"
            />
            <div
              aria-disabled="false"
              aria-label="beat-slider"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="25"
              class="rc-slider-handle"
              role="slider"
              style="height: 0px; border-radius: 0; margin: 0px; bottom: 25%; transform: translateY(+50%);"
              tabindex="0"
            />
            <div
              class="rc-slider-mark"
            />
          </div>
        </div>
        <div
          class="beat"
        >
          <div
            class="rc-slider rc-slider-vertical"
          >
            <div
              class="rc-slider-rail"
              style="background-color: rgb(241, 250, 238); width: 100%; height: 100%; border-radius: 0;"
            />
            <div
              class="rc-slider-track"
              style="background-color: rgb(168, 218, 220); width: 100%; height: 50%; left: 0px; border-radius: 0; bottom: 0%;"
            />
            <div
              class="rc-slider-step"
            />
            <div
              aria-disabled="false"
              aria-label="beat-slider"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="50"
              class="rc-slider-handle"
              role="slider"
              style="height: 0px; border-radius: 0; margin: 0px; bottom: 50%; transform: translateY(+50%);"
              tabindex="0"
            />
            <div
              class="rc-slider-mark"
            />
          </div>
        </div>
      </div>
      <div
        class="flexCenter right"
      >
        <button
          aria-label="add-beat"
        >
          + beat
        </button>
        <button
          aria-label="remove-beat"
        >
          - beat
        </button>
      </div>
    </div>
    <div
      class="row"
    >
      <button
        aria-label="start-stop"
        style="width: 100%;"
      >
        Start
      </button>
    </div>
  </div>
</DocumentFragment>
`;
